lightning/src/util/persist.rs

```diff
@@ -243,10 +250,886 @@ where
+pub struct MonitorUpdatingPersister<K: Deref, L: Deref, ES: Deref + Clone, SP: Deref + Clone>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+	ES::Target: EntropySource + Sized,
+	SP::Target: SignerProvider + Sized,
+{
+	kv_store: K,
+	logger: L,
+	maximum_pending_updates: u64,
+	entropy_source: ES,
+	signer_provider: SP,
+}
+
+#[allow(dead_code)]
+impl<K: Deref, L: Deref, ES: Deref + Clone, SP: Deref + Clone>
+	MonitorUpdatingPersister<K, L, ES, SP>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+	ES::Target: EntropySource + Sized,
+	SP::Target: SignerProvider + Sized,
+{
+	/// Construct a new MonitorUpdatingPersister (MUP).
+	///
+	/// The `maximum_pending_updates` parameter controls how many updates may be stored before a MUP
+	/// consolidates updates by writing a full monitor instead. Note that consolidation will
+	/// frequently occur with fewer updates, this number is merely the maximum that may be stored.
+	pub fn new(
+		kv_store: K,
+		logger: L,
+		maximum_pending_updates: u64,
+		entropy_source: ES,
+		signer_provider: SP,
+	) -> Self
+	where
+		ES::Target: EntropySource + Sized,
+		SP::Target: SignerProvider + Sized,
+	{
+		MonitorUpdatingPersister {
+			kv_store,
+			logger,
+			maximum_pending_updates,
+			entropy_source,
+			signer_provider,
+		}
+	}
+
+	/// Reads channel monitors, along with any stored updates for them.
+	///
+	/// # EXTREMELY IMPORTANT
+	/// It is **EXTREMELY IMPORTANT** that your [`KVStore`] implementation uses the
+	/// [`io::ErrorKind::NotFound`] variant correctly: that is, when a file is not found, and _only_
+	/// in that circumstance (not when there is really a permissions error, for example). This is
+	/// because this function does not list updates. Instead, it lists all monitors, and then using
+	/// their stored `update_id`, synthesizes update storage keys, and tries them in sequence until
+	/// one is not found. All _other_ errors will be bubbled up in this function's [`io::Result`].
+	pub fn read_channel_monitors_with_updates<B: Deref, F: Deref + Clone>(
+		&self,
+		broadcaster: &B,
+		fee_estimator: F,
+	) -> io::Result<
+		Vec<(
+			BlockHash,
+			ChannelMonitor<<SP::Target as SignerProvider>::Signer>,
+		)>,
+	>
+	where
+		ES::Target: EntropySource + Sized,
+		SP::Target: SignerProvider + Sized,
+		B::Target: BroadcasterInterface,
+		F::Target: FeeEstimator,
+	{
+		let mut res = Vec::new();
+		// for each monitor...
+		let monitor_list = self.kv_store.list(
+			CHANNEL_MONITOR_PERSISTENCE_NAMESPACE,
+			CHANNEL_MONITOR_PERSISTENCE_SUB_NAMESPACE,
+		)?;
+		for maybe_monitor_name in monitor_list.into_iter().map(MonitorName::new) {
+			let monitor_name = maybe_monitor_name?;
+			// ...parse the monitor
+			let (bh, monitor) = self.read_monitor(&monitor_name)?;
+			// ...parse and apply the updates with an id higher than the monitor.
+			let mut update_id = monitor.get_latest_update_id();
+			loop {
+				update_id = update_id.saturating_add(1);
```

Comment:
Thanks, `checked_add` is perfect here. I can also bound the loop with that by breaking if it overflows. At some point before, I don't think the `update_persisted_channel` forced writing a whole monitor if the update was `CLOSED_CHANNEL_UPDATE_ID`, so I was concerned about needing to apply an update to a monitor at that height more than once (I think that is a thing that can happen?). Now that doesn't happen so this is the way.

Reply:
I guess you can only get here if you somehow had an update at `u64::MAX`, which is now not a designed behavior of the MUP, so it is an error, in fact. I will fix.

Reply:
Sorry, I had to revert this. When you open a monitor for a closed channel, you _start_ at u64::MAX, so erroring here breaks that (and that is definitely something we need to do).

I think if this is really important to have as an error, it will also need a carve-out for the first iteration being u64::MAX. Do you think that would be an improvement?